# Hands-on: Training and deploying Question Answering with BERT

Pre-trained language representations have been shown to improve many downstream NLP tasks such as question answering, and natural language inference. Devlin, Jacob, et al proposed BERT [1] (Bidirectional Encoder Representations from Transformers), which fine-tunes deep bidirectional representations on a wide range of tasks with minimal task-specific parameters, and obtained state- of-the-art results.

In this tutorial, we will focus on adapting the BERT model for the question answering task on the SQuAD dataset. Specifically, we will:

- understand how to pre-process the SQuAD dataset to leverage the learnt representation in BERT,
- adapt the BERT model to the question answering task, and
- load a trained model to perform inference on the SQuAD dataset

## Sagemaker configuration

This notebook requires mxnet-cu101 >= 1.6.0b20191102, gluonnlp >= 0.8.1
We can create a sagemaker notebook instance with the lifecycle configuration file: sagemaker-lifecycle.config

In [1]:
# One time script
!bash sagemaker-lifecycle.config && pip install mxnet-cu101 --pre -U

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.12
  latest version: 4.8.0

Please update conda by running

    $ conda update -n base -c defaults conda



# All requested packages already installed.

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.12
  latest version: 4.8.0

Please update conda by running

    $ conda update -n base -c defaults conda



# All requested packages already installed.

Skipping mxnet-cu100mkl as it is not installed.
Skipping mxnet-cu101mkl as it is not installed.
You are using pip version 10.0.1, however version 19.3.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
You are using pip version 10.0.1, however version 19.3.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
     \ 2.9MB 2.5MB/ss     / 1.7MB 6.0MB/s
  Running setup.py bdist_wheel for gluonnlp ... done
  Stored 

Requirement not upgraded as not directly required: chardet<3.1.0,>=3.0.2 in /home/ec2-user/anaconda3/envs/mxnet_p36/lib/python3.6/site-packages (from requests<3,>=2.20.0->mxnet-cu101) (3.0.4)
Requirement not upgraded as not directly required: idna<2.8,>=2.5 in /home/ec2-user/anaconda3/envs/mxnet_p36/lib/python3.6/site-packages (from requests<3,>=2.20.0->mxnet-cu101) (2.6)
Requirement not upgraded as not directly required: certifi>=2017.4.17 in /home/ec2-user/anaconda3/envs/mxnet_p36/lib/python3.6/site-packages (from requests<3,>=2.20.0->mxnet-cu101) (2019.11.28)
You are using pip version 10.0.1, however version 19.3.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.


In [2]:
!pip list | grep mxnet
!pip list | grep gluonnlp

keras-mxnet                        2.2.4.2       
mxnet-cu101                        1.6.0b20191122
mxnet-model-server                 1.0.5         
You are using pip version 10.0.1, however version 19.3.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
gluonnlp                           0.9.0.dev0    
You are using pip version 10.0.1, however version 19.3.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.


## Load MXNet and GluonNLP

We first import the libraries:

In [3]:
import argparse, collections, time, logging
import json
import os
import io
import copy
import random
import warnings

import numpy as np
import gluonnlp as nlp
import mxnet as mx
import bert
import qa_utils

from gluonnlp.data import SQuAD
from bert.model.qa import BertForQALoss, BertForQA
from bert.data.qa import SQuADTransform, preprocess_dataset
from bert.bert_qa_evaluate import get_F1_EM, predict, PredResult

# Hyperparameters
parser = argparse.ArgumentParser('BERT finetuning')
parser.add_argument('--epochs', type=int, default=3)
parser.add_argument('--batch_size', default=32)
parser.add_argument('--num_epochs', default=1)
parser.add_argument('--lr', default=5e-5)


args = parser.parse_args([])


epochs = args.epochs
batch_size = args.batch_size
num_epochs = args.num_epochs
lr = args.lr

# output_dir = args.output_dir
# if not os.path.exists(output_dir):
#     os.mkdir(output_dir)
# test_batch_size = args.test_batch_size
# optimizer = args.optimizer
# accumulate = args.accumulate
# warmup_ratio = args.warmup_ratio
# log_interval = args.log_interval
# max_seq_length = args.max_seq_length
# doc_stride = args.doc_stride
# max_query_length = args.max_query_length
# n_best_size = args.n_best_size

## Inspect the SQuAD Dataset

Then we take a look at the Stanford Question Answering Dataset (SQuAD). The dataset can be downloaded using the `nlp.data.SQuAD` API. In this tutorial, we create a small dataset with 3 samples from the SQuAD dataset for demonstration purpose.

The question answering task on the SQuAD dataset is setup the following way. For each sample in the dataset, a context is provided. The context is usually a long paragraph which contains lots of information. Then a question asked based on the context. The goal is to find the text span in the context that answers the question in the sample.

In [4]:
full_data = nlp.data.SQuAD(segment='dev', version='1.1')
# loading a subset of the dev set of SQuAD
num_target_samples = 3
target_samples = [full_data[i] for i in range(num_target_samples)]
dataset = mx.gluon.data.SimpleDataset(target_samples)
print('Number of samples in the created dataset subsampled from SQuAD = %d'%len(dataset))

Number of samples in the created dataset subsampled from SQuAD = 3


In [42]:
target_samples[0]

(0,
 '56be4db0acb8001400a502ec',
 'Which NFL team represented the AFC at Super Bowl 50?',
 'Super Bowl 50 was an American football game to determine the champion of the National Football League (NFL) for the 2015 season. The American Football Conference (AFC) champion Denver Broncos defeated the National Football Conference (NFC) champion Carolina Panthers 24–10 to earn their third Super Bowl title. The game was played on February 7, 2016, at Levi\'s Stadium in the San Francisco Bay Area at Santa Clara, California. As this was the 50th Super Bowl, the league emphasized the "golden anniversary" with various gold-themed initiatives, as well as temporarily suspending the tradition of naming each Super Bowl game with Roman numerals (under which the game would have been known as "Super Bowl L"), so that the logo could prominently feature the Arabic numerals 50.',
 ['Denver Broncos', 'Denver Broncos', 'Denver Broncos'],
 [177, 177, 177])

algo-1-s5jwv_1  | 2020-01-10 21:14:22,976 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Listening on port: /home/model-server/tmp/.mms.sock.9003
algo-1-s5jwv_1  | 2020-01-10 21:14:22,976 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - [PID]932
algo-1-s5jwv_1  | 2020-01-10 21:14:22,976 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - MXNet worker started.
algo-1-s5jwv_1  | 2020-01-10 21:14:22,977 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Python runtime: 3.6.8
algo-1-s5jwv_1  | 2020-01-10 21:14:22,977 [INFO ] W-9003-model com.amazonaws.ml.mms.wlm.WorkerThread - Connecting to: /home/model-server/tmp/.mms.sock.9003
algo-1-s5jwv_1  | 2020-01-10 21:14:22,977 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Connection accepted: /home/model-server/tmp/.mms.sock.9003.
algo-1-s5jwv_1  | 2020-01-10 21:14:24,746 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycl

algo-1-s5jwv_1  | 2020-01-10 21:14:26,956 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Listening on port: /home/model-server/tmp/.mms.sock.9002
algo-1-s5jwv_1  | 2020-01-10 21:14:26,957 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - [PID]966
algo-1-s5jwv_1  | 2020-01-10 21:14:26,957 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - MXNet worker started.
algo-1-s5jwv_1  | 2020-01-10 21:14:26,957 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Python runtime: 3.6.8
algo-1-s5jwv_1  | 2020-01-10 21:14:26,957 [INFO ] W-9002-model com.amazonaws.ml.mms.wlm.WorkerThread - Connecting to: /home/model-server/tmp/.mms.sock.9002
algo-1-s5jwv_1  | 2020-01-10 21:14:26,958 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Connection accepted: /home/model-server/tmp/.mms.sock.9002.
algo-1-s5jwv_1  | 2020-01-10 21:14:27,009 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycl

algo-1-s5jwv_1  | 2020-01-10 21:14:29,189 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Backend worker process die.
algo-1-s5jwv_1  | 2020-01-10 21:14:29,189 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Traceback (most recent call last):
algo-1-s5jwv_1  | 2020-01-10 21:14:29,189 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 172, in <module>
algo-1-s5jwv_1  | 2020-01-10 21:14:29,189 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     worker.run_server()
algo-1-s5jwv_1  | 2020-01-10 21:14:29,189 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 150, in run_server
algo-1-s5jwv_1  | 2020-01-10 21:14:29,190 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     self.handle_connection(cl_socke

algo-1-s5jwv_1  | 2020-01-10 21:14:29,489 [INFO ] W-9006-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Backend worker process die.
algo-1-s5jwv_1  | 2020-01-10 21:14:29,489 [INFO ] W-9006-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Traceback (most recent call last):
algo-1-s5jwv_1  | 2020-01-10 21:14:29,489 [INFO ] W-9006-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 172, in <module>
algo-1-s5jwv_1  | 2020-01-10 21:14:29,489 [INFO ] W-9006-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     worker.run_server()
algo-1-s5jwv_1  | 2020-01-10 21:14:29,489 [INFO ] W-9006-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 150, in run_server
algo-1-s5jwv_1  | 2020-01-10 21:14:29,489 [INFO ] W-9006-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     self.handle_connection(cl_socke

In [46]:
dataset[1]

(1,
 '56be4db0acb8001400a502ed',
 'Which NFL team represented the NFC at Super Bowl 50?',
 'Super Bowl 50 was an American football game to determine the champion of the National Football League (NFL) for the 2015 season. The American Football Conference (AFC) champion Denver Broncos defeated the National Football Conference (NFC) champion Carolina Panthers 24–10 to earn their third Super Bowl title. The game was played on February 7, 2016, at Levi\'s Stadium in the San Francisco Bay Area at Santa Clara, California. As this was the 50th Super Bowl, the league emphasized the "golden anniversary" with various gold-themed initiatives, as well as temporarily suspending the tradition of naming each Super Bowl game with Roman numerals (under which the game would have been known as "Super Bowl L"), so that the logo could prominently feature the Arabic numerals 50.',
 ['Carolina Panthers', 'Carolina Panthers', 'Carolina Panthers'],
 [249, 249, 249])

Let's take a look at a sample from the dataset. In this sample, the question is about the location of the game, with a description about the Super Bowl 50 game as the context. Note that three different answer spans are correct for this question, and they start from index 403, 355 and 355 in the context respectively.

In [5]:
sample = dataset[2]

context_idx = 3

print('\nContext:\n')
print(sample[context_idx])


Context:

Super Bowl 50 was an American football game to determine the champion of the National Football League (NFL) for the 2015 season. The American Football Conference (AFC) champion Denver Broncos defeated the National Football Conference (NFC) champion Carolina Panthers 24–10 to earn their third Super Bowl title. The game was played on February 7, 2016, at Levi's Stadium in the San Francisco Bay Area at Santa Clara, California. As this was the 50th Super Bowl, the league emphasized the "golden anniversary" with various gold-themed initiatives, as well as temporarily suspending the tradition of naming each Super Bowl game with Roman numerals (under which the game would have been known as "Super Bowl L"), so that the logo could prominently feature the Arabic numerals 50.


In [6]:
print(dataset[2])

(2, '56be4db0acb8001400a502ee', 'Where did Super Bowl 50 take place?', 'Super Bowl 50 was an American football game to determine the champion of the National Football League (NFL) for the 2015 season. The American Football Conference (AFC) champion Denver Broncos defeated the National Football Conference (NFC) champion Carolina Panthers 24–10 to earn their third Super Bowl title. The game was played on February 7, 2016, at Levi\'s Stadium in the San Francisco Bay Area at Santa Clara, California. As this was the 50th Super Bowl, the league emphasized the "golden anniversary" with various gold-themed initiatives, as well as temporarily suspending the tradition of naming each Super Bowl game with Roman numerals (under which the game would have been known as "Super Bowl L"), so that the logo could prominently feature the Arabic numerals 50.', ['Santa Clara, California', "Levi's Stadium", "Levi's Stadium in the San Francisco Bay Area at Santa Clara, California."], [403, 355, 355])


In [7]:
print(dataset[1])

(1, '56be4db0acb8001400a502ed', 'Which NFL team represented the NFC at Super Bowl 50?', 'Super Bowl 50 was an American football game to determine the champion of the National Football League (NFL) for the 2015 season. The American Football Conference (AFC) champion Denver Broncos defeated the National Football Conference (NFC) champion Carolina Panthers 24–10 to earn their third Super Bowl title. The game was played on February 7, 2016, at Levi\'s Stadium in the San Francisco Bay Area at Santa Clara, California. As this was the 50th Super Bowl, the league emphasized the "golden anniversary" with various gold-themed initiatives, as well as temporarily suspending the tradition of naming each Super Bowl game with Roman numerals (under which the game would have been known as "Super Bowl L"), so that the logo could prominently feature the Arabic numerals 50.', ['Carolina Panthers', 'Carolina Panthers', 'Carolina Panthers'], [249, 249, 249])


In [8]:
print(dataset[0])

(0, '56be4db0acb8001400a502ec', 'Which NFL team represented the AFC at Super Bowl 50?', 'Super Bowl 50 was an American football game to determine the champion of the National Football League (NFL) for the 2015 season. The American Football Conference (AFC) champion Denver Broncos defeated the National Football Conference (NFC) champion Carolina Panthers 24–10 to earn their third Super Bowl title. The game was played on February 7, 2016, at Levi\'s Stadium in the San Francisco Bay Area at Santa Clara, California. As this was the 50th Super Bowl, the league emphasized the "golden anniversary" with various gold-themed initiatives, as well as temporarily suspending the tradition of naming each Super Bowl game with Roman numerals (under which the game would have been known as "Super Bowl L"), so that the logo could prominently feature the Arabic numerals 50.', ['Denver Broncos', 'Denver Broncos', 'Denver Broncos'], [177, 177, 177])


In [9]:
type(dataset)

mxnet.gluon.data.dataset.SimpleDataset

In [10]:
question_idx = 2
answer_idx = 4
answer_pos_idx = 5

print("\nQuestion")
print(sample[question_idx])
print("\nCorrect Answer Spans")
print(sample[answer_idx])
print("\nAnswer Span Start Indices:")
print(sample[answer_pos_idx])


Question
Where did Super Bowl 50 take place?

Correct Answer Spans
['Santa Clara, California', "Levi's Stadium", "Levi's Stadium in the San Francisco Bay Area at Santa Clara, California."]

Answer Span Start Indices:
[403, 355, 355]


## Data Pre-processing for QA with BERT

Recall that during BERT pre-training, it takes a sentence pair as the input, separated by the 'SEP' special token. For SQuAD, we can feed the context-question pair as the sentence pair input. To use BERT to predict the starting and ending span of the answer, we can add a classification layer for each token in the context texts, to predict if a token is the start or the end of the answer span. 

![qa](natural_language_understanding/qa.png)

In the next few code blocks, we will work on pre-processing the samples in the SQuAD dataset in the desired format with these special separators. 


### Get Pre-trained BERT Model

First, let's use the *get_model* API in GluonNLP to get the model definition for BERT, and the vocabulary used for the BERT model. Note that we discard the pooler and classifier layers used for the next sentence prediction task, as well as the decoder layers for the masked language model task during the BERT pre-training phase. These layers are not useful for predicting the starting and ending indices of the answer span.

The list of pre-trained BERT models available in GluonNLP can be found [here](http://gluon-nlp.mxnet.io/model_zoo/bert/index.html).

In [11]:
bert_model, vocab = nlp.model.get_model('bert_12_768_12',
                                        dataset_name='book_corpus_wiki_en_uncased',
                                        use_classifier=False,
                                        use_decoder=False,
                                        use_pooler=False,
                                        pretrained=False)

Vocab file is not found. Downloading.


Note that there are several special tokens in the vocabulary for BERT. In particular, the `[SEP]` token is used for separating the sentence pairs, and the `[CLS]` token is added at the beginning of the sentence pairs. They will be used to pre-process the SQuAD dataset later.

In [12]:
print(vocab)

Vocab(size=30522, unk="[UNK]", reserved="['[CLS]', '[SEP]', '[MASK]', '[PAD]']")


### Tokenization

The second step is to process the samples using the same tokenizer used for BERT, which is provided as the `BERTTokenizer` API in GluonNLP. Note that instead of word level and character level representation, BERT uses subwords to represent a word, separated `##`. 

In the following example, the word `suspending` is tokenized as two subwords (`suspend` and `##ing`), and `numerals` is tokenized as three subwords (`nu`, `##meral`, `##s`).

In [13]:
tokenizer = nlp.data.BERTTokenizer(vocab=vocab, lower=True)

tokenizer("as well as temporarily suspending the tradition of naming each Super Bowl game with Roman numerals")

['as',
 'well',
 'as',
 'temporarily',
 'suspend',
 '##ing',
 'the',
 'tradition',
 'of',
 'naming',
 'each',
 'super',
 'bowl',
 'game',
 'with',
 'roman',
 'nu',
 '##meral',
 '##s']

### Sentence Pair Composition

With the tokenizer inplace, we are ready to process the question-context texts and compose sentence pairs. The functionality is available via the `SQuADTransform` API. 

In [14]:
transform = bert.data.qa.SQuADTransform(tokenizer, is_pad=False, is_training=False, do_lookup=False)
dev_data_transform, _ = bert.data.qa.preprocess_dataset(dataset, transform)
logging.info('The number of examples after preprocessing:{}'.format(len(dev_data_transform)))

Done! Transform dataset costs 0.17 seconds.


Let's take a look at the sample after the transformation:

In [15]:
## test

print(len(sample))
print(type(sample))
print(sample) ## (example_id, [question, content], ques_cont_token_types, valid_length, _, _)


6
<class 'tuple'>
(2, '56be4db0acb8001400a502ee', 'Where did Super Bowl 50 take place?', 'Super Bowl 50 was an American football game to determine the champion of the National Football League (NFL) for the 2015 season. The American Football Conference (AFC) champion Denver Broncos defeated the National Football Conference (NFC) champion Carolina Panthers 24–10 to earn their third Super Bowl title. The game was played on February 7, 2016, at Levi\'s Stadium in the San Francisco Bay Area at Santa Clara, California. As this was the 50th Super Bowl, the league emphasized the "golden anniversary" with various gold-themed initiatives, as well as temporarily suspending the tradition of naming each Super Bowl game with Roman numerals (under which the game would have been known as "Super Bowl L"), so that the logo could prominently feature the Arabic numerals 50.', ['Santa Clara, California', "Levi's Stadium", "Levi's Stadium in the San Francisco Bay Area at Santa Clara, California."], [403, 35

In [16]:
sample = dev_data_transform[2]
print('\nsegment type: \n' + str(sample[2]))
print('\ntext length: ' + str(sample[3]))
print('\nsentence pair: \n' + str(sample[1]))


segment type: 
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]

text length: 168

sentence pair: 
['[CLS]', 'where', 'did', 'super', 'bowl', '50', 'take', 'place', '?', '[SEP]', 'super', 'bowl', '50', 'was', 'an', 'american', 'football', 'game', 'to', 'determine', 'the', 'champion', 'of', 'the', 'national', 'football', 'league', '(', 'nfl', ')', 'for', 'the', '2015', 'season', '.', 'the', 'american', 'football', 'conference', '(', 'afc', ')', 'champion', 'denver', 'broncos', 'defeated', 'the', 'national', 'football', 'conference', '(', 

### Vocabulary Lookup

Finally, we convert the transformed texts to subword indices, which are used to contructor NDArrays as the inputs to the model.

In [17]:
def vocab_lookup(example_id, subwords, type_ids, length, start, end):
    indices = vocab[subwords]
    return example_id, indices, type_ids, length, start, end

dev_data_transform = dev_data_transform.transform(vocab_lookup, lazy=False)
print(dev_data_transform[2][1])

[2, 2073, 2106, 3565, 4605, 2753, 2202, 2173, 1029, 3, 3565, 4605, 2753, 2001, 2019, 2137, 2374, 2208, 2000, 5646, 1996, 3410, 1997, 1996, 2120, 2374, 2223, 1006, 5088, 1007, 2005, 1996, 2325, 2161, 1012, 1996, 2137, 2374, 3034, 1006, 10511, 1007, 3410, 7573, 14169, 3249, 1996, 2120, 2374, 3034, 1006, 22309, 1007, 3410, 3792, 12915, 2484, 1516, 2184, 2000, 7796, 2037, 2353, 3565, 4605, 2516, 1012, 1996, 2208, 2001, 2209, 2006, 2337, 1021, 1010, 2355, 1010, 2012, 11902, 1005, 1055, 3346, 1999, 1996, 2624, 3799, 3016, 2181, 2012, 4203, 10254, 1010, 2662, 1012, 2004, 2023, 2001, 1996, 12951, 3565, 4605, 1010, 1996, 2223, 13155, 1996, 1000, 3585, 5315, 1000, 2007, 2536, 2751, 1011, 11773, 11107, 1010, 2004, 2092, 2004, 8184, 28324, 2075, 1996, 4535, 1997, 10324, 2169, 3565, 4605, 2208, 2007, 3142, 16371, 28990, 2015, 1006, 2104, 2029, 1996, 2208, 2052, 2031, 2042, 2124, 2004, 1000, 3565, 4605, 1048, 1000, 1007, 1010, 2061, 2008, 1996, 8154, 2071, 14500, 3444, 1996, 5640, 16371, 28990, 2015

## Model Definition

After the data is processed, we can define the model that uses the representation produced by BERT for predicting the starting and ending positions of the answer span.

We download a BERT model trained on the SQuAD dataset, prepare the dataloader.

In [18]:
net = BertForQA(bert_model)
ctx = mx.gpu(0)
ckpt = qa_utils.download_qa_ckpt()
net.load_parameters(ckpt, ctx=ctx)

batch_size = 1
dev_dataloader = mx.gluon.data.DataLoader(
    dev_data_transform, batch_size=batch_size, shuffle=False)

Downloaded checkpoint to ./temp/bert_qa-7eb11865.params


In [19]:
## test

for data in dev_dataloader:
    example_ids, inputs, token_types, valid_length, _, _ = data
    print(valid_length)


[171]
<NDArray 1 @cpu(0)>

[171]
<NDArray 1 @cpu(0)>

[168]
<NDArray 1 @cpu(0)>


In [20]:
## test

print(dev_dataloader)

In [21]:
all_results = collections.defaultdict(list)

total_num = 0
for data in dev_dataloader:
    example_ids, inputs, token_types, valid_length, _, _ = data
    total_num += len(inputs)
    batch_size = inputs.shape[0]
    output = net(inputs.astype('float32').as_in_context(ctx),
                               token_types.astype('float32').as_in_context(ctx),
                               valid_length.astype('float32').as_in_context(ctx))
    pred_start, pred_end = mx.nd.split(output, axis=2, num_outputs=2)
    example_ids = example_ids.asnumpy().tolist()
    pred_start = pred_start.reshape(batch_size, -1).asnumpy()
    pred_end = pred_end.reshape(batch_size, -1).asnumpy()
    
    for example_id, start, end in zip(example_ids, pred_start, pred_end):
        all_results[example_id].append(PredResult(start=start, end=end))

In [22]:
qa_utils.predict(dataset, all_results, vocab)


Context: Super Bowl 50 was an American football game to determine the champion of the National Football League (NFL) for the 2015 season. The American Football Conference (AFC) champion Denver Broncos defeated the National Football Conference (NFC) champion Carolina Panthers 24–10 to earn their third Super Bowl title. The game was played on February 7, 2016, at Levi's Stadium in the San Francisco Bay Area at Santa Clara, California. As this was the 50th Super Bowl, the league emphasized the "golden anniversary" with various gold-themed initiatives, as well as temporarily suspending the tradition of naming each Super Bowl game with Roman numerals (under which the game would have been known as "Super Bowl L"), so that the logo could prominently feature the Arabic numerals 50.

Question: which nfl team represented the afc at super bowl 50 ?

Top predictions: 
99.36% 	 Denver Broncos
0.23% 	 The American Football Conference (AFC) champion Denver Broncos
0.20% 	 Broncos


Context: Super Bo

### Let's Train the Model

Now we can put all the pieces together, and start fine-tuning the model with a few epochs.

In [23]:
# net = BertForQA(bert=bert_model)
# nlp.utils.load_parameters(net, pretrained_bert_parameters, ctx=ctx,
#                           ignore_extra=True, cast_dtype=True)
net.span_classifier.initialize(init=mx.init.Normal(0.02), ctx=ctx)
net.hybridize(static_alloc=True)

loss_function = BertForQALoss()
loss_function.hybridize(static_alloc=True)

/home/ec2-user/anaconda3/envs/mxnet_p36/lib/python3.6/site-packages/mxnet/gluon/parameter.py:887: UserWarning: Parameter 'bertforqa0_dense0_weight' is already initialized, ignoring. Set force_reinit=True to re-initialize.
  v.initialize(None, ctx, init, force_reinit=force_reinit)
/home/ec2-user/anaconda3/envs/mxnet_p36/lib/python3.6/site-packages/mxnet/gluon/parameter.py:887: UserWarning: Parameter 'bertforqa0_dense0_bias' is already initialized, ignoring. Set force_reinit=True to re-initialize.
  v.initialize(None, ctx, init, force_reinit=force_reinit)


## Deploy on SageMaker

1. Saving the model parameters
2. Preparing functions for inference 
3. Building a docker container with dependencies installed
4. Launching a serving end-point with SageMaker SDK

### 1. Saving the model parameters

In [24]:
## save parameters, model definition and vocabulary in a zip file

# net.export('checkpoint')

output_dir = "model_outputs"
if not os.path.exists(output_dir):
    os.mkdir(output_dir)

with open(output_dir+'/vocab.json', 'w') as f:
    f.write(vocab.to_json())
import tarfile
with tarfile.open(output_dir+"/model.tar.gz", "w:gz") as tar:
    tar.add("./temp/bert_qa-7eb11865.params")
#     tar.add("output_dir/checkpoint-0000.params") 
#     tar.add("output_dir/checkpoint-symbol.json") ???
    tar.add(output_dir+"/vocab.json")

### 2. Preparing functions for inference

Two functions: 
1. model_fn() to load model parameters
2. transform_fn() to run model inference given an input

In [57]:
## test

test_example_0 = ['Which NFL team represented the AFC at Super Bowl 50?',
 'Super Bowl 50 was an American football game to determine the champion of the National Football League (NFL) for the 2015 season. The American Football Conference (AFC) champion Denver Broncos defeated the National Football Conference (NFC) champion Carolina Panthers 24–10 to earn their third Super Bowl title. The game was played on February 7, 2016, at Levi\'s Stadium in the San Francisco Bay Area at Santa Clara, California. As this was the 50th Super Bowl, the league emphasized the "golden anniversary" with various gold-themed initiatives, as well as temporarily suspending the tradition of naming each Super Bowl game with Roman numerals (under which the game would have been known as "Super Bowl L"), so that the logo could prominently feature the Arabic numerals 50.']

test_example_1 = ['Where did Super Bowl 50 take place?',
 'Super Bowl 50 was an American football game to determine the champion of the National Football League (NFL) for the 2015 season. The American Football Conference (AFC) champion Denver Broncos defeated the National Football Conference (NFC) champion Carolina Panthers 24–10 to earn their third Super Bowl title. The game was played on February 7, 2016, at Levi\'s Stadium in the San Francisco Bay Area at Santa Clara, California. As this was the 50th Super Bowl, the league emphasized the "golden anniversary" with various gold-themed initiatives, as well as temporarily suspending the tradition of naming each Super Bowl game with Roman numerals (under which the game would have been known as "Super Bowl L"), so that the logo could prominently feature the Arabic numerals 50.']

my_test_examples = [test_example_0, test_example_1]


In [82]:
%%writefile serve.py
import collections, json, logging, warnings
import gluonnlp as nlp
import mxnet as mx
from mxnet.gluon import Block, nn
print('about to import')
import bert
# from bert.data.qa import preprocess_dataset, SQuADTransform
# from bert.bert_qa_evaluate import PredResult, predict

# import multiprocessing as mp
# from functools import partial

# class BertForQA(Block):
#     """Model for SQuAD task with BERT.
#     The model feeds token ids and token type ids into BERT to get the
#     pooled BERT sequence representation, then apply a Dense layer for QA task.
#     Parameters
#     ----------
#     bert: BERTModel
#         Bidirectional encoder with transformer.
#     prefix : str or None
#         See document of `mx.gluon.Block`.
#     params : ParameterDict or None
#         See document of `mx.gluon.Block`.
#     """

#     def __init__(self, bert, prefix=None, params=None):
#         super(BertForQA, self).__init__(prefix=prefix, params=params)
#         self.bert = bert
#         with self.name_scope():
#             self.span_classifier = nn.Dense(units=2, flatten=False)

#     def forward(self, inputs, token_types, valid_length=None):  # pylint: disable=arguments-differ
#         """Generate the unnormalized score for the given the input sequences.
#         Parameters
#         ----------
#         inputs : NDArray, shape (batch_size, seq_length)
#             Input words for the sequences.
#         token_types : NDArray, shape (batch_size, seq_length)
#             Token types for the sequences, used to indicate whether the word belongs to the
#             first sentence or the second one.
#         valid_length : NDArray or None, shape (batch_size,)
#             Valid length of the sequence. This is used to mask the padded tokens.
#         Returns
#         -------
#         outputs : NDArray
#             Shape (batch_size, seq_length, 2)
#         """
#         bert_output = self.bert(inputs, token_types, valid_length)
#         output = self.span_classifier(bert_output)
#         return output
    
# def get_all_results(net, dev_data_transform):
#     all_results = collections.defaultdict(list)
#     dev_dataloader = mx.gluon.data.DataLoader(dev_data_transform, batch_size=1, shuffle=False)
#     for data in dev_dataloader:
#         example_ids, inputs, token_types, valid_length, _, _ = data
#         batch_size = inputs.shape[0]
#         output = net(inputs.astype('float32').as_in_context(ctx),
#                                    token_types.astype('float32').as_in_context(ctx),
#                                    valid_length.astype('float32').as_in_context(ctx))
#         pred_start, pred_end = mx.nd.split(output, axis=2, num_outputs=2)
#         example_ids = example_ids.asnumpy().tolist()
#         pred_start = pred_start.reshape(batch_size, -1).asnumpy()
#         pred_end = pred_end.reshape(batch_size, -1).asnumpy()

#         for example_id, start, end in zip(example_ids, pred_start, pred_end):
#             all_results[example_id].append(PredResult(start=start, end=end))
#     return(all_results)

# def test_example_transform(test_examples):
#     test_examples_tuples = []
#     i = 0
#     for test in test_examples:
#         tup = (i, "", test[0], test[1], [], [])
#         test_examples_tuples.append(tup)
#         i += 1
#     return(test_examples_tuples)


# def model_fn(params_path = "temp/bert_qa-7eb11865.params"):
#     """
#     Load the gluon model. Called once when hosting service starts.
#     :param: model_dir The directory where model files are stored.
#     :return: a Gluon model, and the vocabulary
#     """
#     print('a')
#     return None

# #     bert_model, vocab = nlp.model.get_model('bert_12_768_12',
# #                                         dataset_name='book_corpus_wiki_en_uncased',
# #                                         use_classifier=False,
# #                                         use_decoder=False,
# #                                         use_pooler=False,
# #                                         pretrained=False)
# #     net = BertForQA(bert_model)
# #     net.load_parameters(params_path, ctx=mx.cpu())
# #     print('here')
# #     tokenizer = nlp.data.BERTTokenizer(vocab, lower=True)
# #     print('there')
# #     transform = bert.qa.data.SQuADTransform(tokenizer, is_pad=False, is_training=False, do_lookup=False)
# #     print('there2')

# #     return net, vocab, transform


# def transform_fn(model, data, input_content_type, output_content_type):
#     """
#     Transform a request using the Gluon model. Called once per request.
#     :param model: The Gluon model and the vocab
#     :param dataset: The request payload
    
#         Example:
#         ## (example_id, [question, content], ques_cont_token_types, valid_length, _, _)


#         (2, 
#         '56be4db0acb8001400a502ee', 
#         'Where did Super Bowl 50 take place?', 
        
#         'Super Bowl 50 was an American football game to determine the champion of the National 
#         Football League (NFL) for the 2015 season. The American Football Conference (AFC) 
#         champion Denver Broncos defeated the National Football Conference (NFC) champion 
#         Carolina Panthers 24–10 to earn their third Super Bowl title. The game was played 
#         on February 7, 2016, at Levi\'s Stadium in the San Francisco Bay Area at Santa Clara, 
#         California. As this was the 50th Super Bowl, the league emphasized the "golden 
#         anniversary" with various gold-themed initiatives, as well as temporarily suspending 
#         the tradition of naming each Super Bowl game with Roman numerals (under which the 
#         game would have been known as "Super Bowl L"), so that the logo could prominently 
#         feature the Arabic numerals 50.', 
        
#         ['Santa Clara, California', "Levi's Stadium", "Levi's Stadium 
#         in the San Francisco Bay Area at Santa Clara, California."], 
        
#         [403, 355, 355])
        

        
#     :param input_content_type: The request content type, assume json
#     :param output_content_type: The (desired) response content type, assume json
#     :return: response payload and content type.
#     """
    

#     net, vocab, transform = model
#     test_examples = json.loads(data)
#     my_test_examples_tuples = test_example_transform(test_examples)
#     test_dataset = mx.gluon.data.SimpleDataset(my_test_examples_tuples)
# #     dev_data_transform, _ = preprocess_dataset(dataset, transform)
# #     dev_data_transform = dev_data_transform.transform(vocab_lookup, lazy=False)
#     dev_data_transform = test_dataset.transform(transform._transform)
    
#     all_predictions = collections.defaultdict(list) # collections.OrderedDict()
#     all_results = get_all_results(net, dev_data_transform)
#     for features in dev_data_transform:
#         f_id = features[0].example_id
#         results = all_results[f_id]
    
#         prediction, nbest = predict(
#             features=features,
#             results=results,
#             tokenizer=nlp.data.BERTBasicTokenizer(vocab, lower=True))
        
#         nbest_prediction = [] 
#         for i in range(3):
#             nbest_prediction.append('%.2f%% \t %s'%(nbest[i][1] * 100, nbest[i][0]))
#         all_predictions[f_id] = nbest_prediction
#     response_body = json.dumps(all_predictions)
#     return response_body, output_content_type


Overwriting serve.py


In [83]:
## test
import serve
mymodel = serve.model_fn()
transform_fn(mymodel, my_test_examples)

NameError: name 'bert' is not defined

In [65]:

# ## prepare dataset json
# my_test_examples_tuples = test_example_transform(my_test_examples)
# with open('test_examples_tuples.json', 'w') as f:
#     json.dump(my_test_examples_tuples, f)

# test_dataset[1]



# with open('dataset.json') as data_file:    
#     data = json.load(data_file)

### 3. Building a docker container with dependencies installed

Let's prepare a docker container with all the dependencies required for model inference. Here we build a docker container based on the SageMaker MXNet inference container, and you can find the list of all available inference containers at https://docs.aws.amazon.com/sagemaker/latest/dg/pre-built-containers-frameworks-deep-learning.html

Here we use local mode for demonstration purpose. To deploy on actual instances, you need to login into AWS elastic container registry (ECR) service, and push the container to ECR. 

```
docker build -t $YOUR_EDR_DOCKER_TAG . -f Dockerfile
$(aws ecr get-login --no-include-email --region $YOUR_REGION)
docker push $YOUR_EDR_DOCKER_TAG
```

In [26]:
%%writefile Dockerfile

ARG REGION
FROM 763104351884.dkr.ecr.$REGION.amazonaws.com/mxnet-inference:1.4.1-gpu-py3

RUN pip install --upgrade --user --pre 'mxnet-mkl' 'https://github.com/dmlc/gluon-nlp/tarball/v0.9.x'

RUN pip list | grep mxnet

COPY *.py /opt/ml/model/code/
COPY bert/data/qa.py /opt/ml/model/code/bert/data/
COPY bert/bert_qa_evaluate.py /opt/ml/model/code/bert/

Overwriting Dockerfile


Docker login cmd

In [27]:
!$(aws ecr get-login --no-include-email --region us-east-1 --registry-ids 763104351884)

WARNING! Using --password via the CLI is insecure. Use --password-stdin.
WARNING! Your password will be stored unencrypted in /home/ec2-user/.docker/config.json.
Configure a credential helper to remove this warning. See
https://docs.docker.com/engine/reference/commandline/login/#credentials-store

Login Succeeded


In [28]:
!export REGION=$(wget -qO- http://169.254.169.254/latest/meta-data/placement/availability-zone) &&\
 docker build --no-cache --build-arg REGION=${REGION::-1} -t my-docker:inference . -f Dockerfile

Sending build context to Docker daemon   1.28GB
Step 1/7 : ARG REGION
Step 2/7 : FROM 763104351884.dkr.ecr.$REGION.amazonaws.com/mxnet-inference:1.4.1-gpu-py3
 ---> d9dd4dcfe0c2
Step 3/7 : RUN pip install --upgrade --user --pre 'mxnet-mkl' 'https://github.com/dmlc/gluon-nlp/tarball/v0.9.x'
 ---> Running in 0676dcb80b06
ERROR: mxnet-mkl 1.6.0b20191025 has requirement numpy<2.0.0,>1.16.0, but you'll have numpy 1.14.6 which is incompatible.
  Running setup.py install for gluonnlp: started
    Running setup.py install for gluonnlp: finished with status 'done'
You should consider upgrading via the 'pip install --upgrade pip' command.
Removing intermediate container 0676dcb80b06
 ---> 004e4fd7d2f5
Step 4/7 : RUN pip list | grep mxnet
 ---> Running in e6fd5c91cb07
You should consider upgrading via the 'pip install --upgrade pip' command.
keras-mxnet                       2.2.4.1       
mxnet-cu100mkl                    1.4.1         
mxnet-mkl                         1.6.0b20191025
mxnet-mode

### 4. Launching a serving end-point with SageMaker SDK

We create a MXNet model which can be deployed later, by specifying the docker image, and entry point for the inference code. If serve.py does not work, use dummy_hosting_module.py for debugging purpose. 

In [29]:
import sagemaker
from sagemaker.mxnet.model import MXNetModel
sagemaker_model = MXNetModel(model_data='file:///home/ec2-user/SageMaker/ako2020-bert/tutorial/model_outputs/model.tar.gz',
                             image='my-docker:inference', # docker images
                             role=sagemaker.get_execution_role(), 
                             py_version='py3',            # python version
                             entry_point='serve.py',
                             source_dir='.')

We use 'local' mode to test our deployment code, where the inference happens on the current instance.
If you are ready to deploy the model on a new instance, change the `instance_type` argument to values such as `ml.c4.xlarge`.

Here we use 'local' mode for testing, for real instances use c5.2xlarge, p2.xlarge, etc.

In [30]:
predictor = sagemaker_model.deploy(initial_instance_count=1, instance_type='local')

Attaching to tmpmeyjgvsd_algo-1-s5jwv_1
algo-1-s5jwv_1  | 2020-01-10 20:57:20,647 [INFO ] main com.amazonaws.ml.mms.ModelServer - 
algo-1-s5jwv_1  | MMS Home: /usr/local/lib/python3.6/site-packages
algo-1-s5jwv_1  | Current directory: /
algo-1-s5jwv_1  | Temp directory: /home/model-server/tmp
algo-1-s5jwv_1  | Number of GPUs: 0
algo-1-s5jwv_1  | Number of CPUs: 8
algo-1-s5jwv_1  | Max heap size: 13646 M
algo-1-s5jwv_1  | Python executable: /usr/local/bin/python3.6
algo-1-s5jwv_1  | Config file: /etc/sagemaker-mms.properties
algo-1-s5jwv_1  | Inference address: http://0.0.0.0:8080
algo-1-s5jwv_1  | Management address: http://127.0.0.1:8081
algo-1-s5jwv_1  | Model Store: /.sagemaker/mms/models
algo-1-s5jwv_1  | Initial Models: ALL
algo-1-s5jwv_1  | Log dir: /logs
algo-1-s5jwv_1  | Metrics dir: /logs
algo-1-s5jwv_1  | Netty threads: 0
algo-1-s5jwv_1  | Netty client threads: 0
algo-1-s5jwv_1  | Default workers per model: 8
algo-1-s5jwv_1  | Blacklist Regex: N/A
algo-1-s5jwv_1  | Maximum Re

algo-1-s5jwv_1  | 2020-01-10 20:57:20,901 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Listening on port: /home/model-server/tmp/.mms.sock.9002
algo-1-s5jwv_1  | 2020-01-10 20:57:20,902 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - [PID]77
algo-1-s5jwv_1  | 2020-01-10 20:57:20,902 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - MXNet worker started.
algo-1-s5jwv_1  | 2020-01-10 20:57:20,904 [INFO ] W-9001-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Listening on port: /home/model-server/tmp/.mms.sock.9001
algo-1-s5jwv_1  | 2020-01-10 20:57:20,904 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Python runtime: 3.6.8
algo-1-s5jwv_1  | 2020-01-10 20:57:20,904 [INFO ] W-9001-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - [PID]78
algo-1-s5jwv_1  | 2020-01-10 20:57:20,905 [INFO ] W-9001-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - MXNet worker started.
algo-1-s5jwv_

algo-1-s5jwv_1  | 2020-01-10 20:57:23,736 [INFO ] W-9000-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Backend worker process die.
algo-1-s5jwv_1  | 2020-01-10 20:57:23,736 [INFO ] W-9000-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Traceback (most recent call last):
algo-1-s5jwv_1  | 2020-01-10 20:57:23,736 [INFO ] W-9000-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 172, in <module>
algo-1-s5jwv_1  | 2020-01-10 20:57:23,738 [INFO ] W-9000-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     worker.run_server()
algo-1-s5jwv_1  | 2020-01-10 20:57:23,738 [INFO ] W-9000-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 150, in run_server
algo-1-s5jwv_1  | 2020-01-10 20:57:23,738 [INFO ] W-9000-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     self.handle_connection(cl_socke

algo-1-s5jwv_1  | 2020-01-10 20:57:24,939 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Listening on port: /home/model-server/tmp/.mms.sock.9002
algo-1-s5jwv_1  | 2020-01-10 20:57:24,944 [INFO ] W-9000-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Listening on port: /home/model-server/tmp/.mms.sock.9000
algo-1-s5jwv_1  | 2020-01-10 20:57:24,945 [INFO ] W-9000-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - [PID]156
algo-1-s5jwv_1  | 2020-01-10 20:57:24,945 [INFO ] W-9000-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - MXNet worker started.
algo-1-s5jwv_1  | 2020-01-10 20:57:24,945 [INFO ] W-9000-model com.amazonaws.ml.mms.wlm.WorkerThread - Connecting to: /home/model-server/tmp/.mms.sock.9000
algo-1-s5jwv_1  | 2020-01-10 20:57:24,946 [INFO ] W-9000-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Python runtime: 3.6.8
algo-1-s5jwv_1  | 2020-01-10 20:57:24,946 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -

In [31]:
output = predictor.predict(dataset)  
print('\nPrediction output: {}\n\n'.format(output))

TypeError: Object of type 'SimpleDataset' is not JSON serializable

algo-1-s5jwv_1  | 2020-01-10 20:57:27,803 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Backend worker process die.
algo-1-s5jwv_1  | 2020-01-10 20:57:27,803 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Traceback (most recent call last):
algo-1-s5jwv_1  | 2020-01-10 20:57:27,803 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 172, in <module>
algo-1-s5jwv_1  | 2020-01-10 20:57:27,804 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     worker.run_server()
algo-1-s5jwv_1  | 2020-01-10 20:57:27,804 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 150, in run_server
algo-1-s5jwv_1  | 2020-01-10 20:57:27,804 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     self.handle_connection(cl_socke

algo-1-s5jwv_1  | 2020-01-10 20:57:28,932 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Listening on port: /home/model-server/tmp/.mms.sock.9003
algo-1-s5jwv_1  | 2020-01-10 20:57:28,933 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - [PID]221
algo-1-s5jwv_1  | 2020-01-10 20:57:28,933 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - MXNet worker started.
algo-1-s5jwv_1  | 2020-01-10 20:57:28,933 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Python runtime: 3.6.8
algo-1-s5jwv_1  | 2020-01-10 20:57:28,933 [INFO ] W-9003-model com.amazonaws.ml.mms.wlm.WorkerThread - Connecting to: /home/model-server/tmp/.mms.sock.9003
algo-1-s5jwv_1  | 2020-01-10 20:57:28,935 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Connection accepted: /home/model-server/tmp/.mms.sock.9003.
algo-1-s5jwv_1  | 2020-01-10 20:57:28,940 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycl

algo-1-s5jwv_1  | 2020-01-10 20:57:31,893 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Backend worker process die.
algo-1-s5jwv_1  | 2020-01-10 20:57:31,893 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Traceback (most recent call last):
algo-1-s5jwv_1  | 2020-01-10 20:57:31,893 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 172, in <module>
algo-1-s5jwv_1  | 2020-01-10 20:57:31,893 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     worker.run_server()
algo-1-s5jwv_1  | 2020-01-10 20:57:31,893 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 150, in run_server
algo-1-s5jwv_1  | 2020-01-10 20:57:31,893 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     self.handle_connection(cl_socke

algo-1-s5jwv_1  | 2020-01-10 20:57:33,797 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Listening on port: /home/model-server/tmp/.mms.sock.9003
algo-1-s5jwv_1  | 2020-01-10 20:57:33,797 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - [PID]276
algo-1-s5jwv_1  | 2020-01-10 20:57:33,797 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - MXNet worker started.
algo-1-s5jwv_1  | 2020-01-10 20:57:33,797 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Python runtime: 3.6.8
algo-1-s5jwv_1  | 2020-01-10 20:57:33,797 [INFO ] W-9003-model com.amazonaws.ml.mms.wlm.WorkerThread - Connecting to: /home/model-server/tmp/.mms.sock.9003
algo-1-s5jwv_1  | 2020-01-10 20:57:33,798 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Connection accepted: /home/model-server/tmp/.mms.sock.9003.
algo-1-s5jwv_1  | 2020-01-10 20:57:33,877 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycl

algo-1-s5jwv_1  | 2020-01-10 20:57:36,770 [INFO ] W-9005-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Backend worker process die.
algo-1-s5jwv_1  | 2020-01-10 20:57:36,770 [INFO ] W-9005-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Traceback (most recent call last):
algo-1-s5jwv_1  | 2020-01-10 20:57:36,770 [INFO ] W-9005-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 172, in <module>
algo-1-s5jwv_1  | 2020-01-10 20:57:36,770 [INFO ] W-9005-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     worker.run_server()
algo-1-s5jwv_1  | 2020-01-10 20:57:36,770 [INFO ] W-9005-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 150, in run_server
algo-1-s5jwv_1  | 2020-01-10 20:57:36,771 [INFO ] W-9005-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     self.handle_connection(cl_socke

algo-1-s5jwv_1  | 2020-01-10 20:57:39,661 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Listening on port: /home/model-server/tmp/.mms.sock.9003
algo-1-s5jwv_1  | 2020-01-10 20:57:39,661 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - [PID]331
algo-1-s5jwv_1  | 2020-01-10 20:57:39,662 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - MXNet worker started.
algo-1-s5jwv_1  | 2020-01-10 20:57:39,662 [INFO ] W-9003-model com.amazonaws.ml.mms.wlm.WorkerThread - Connecting to: /home/model-server/tmp/.mms.sock.9003
algo-1-s5jwv_1  | 2020-01-10 20:57:39,662 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Python runtime: 3.6.8
algo-1-s5jwv_1  | 2020-01-10 20:57:39,662 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Connection accepted: /home/model-server/tmp/.mms.sock.9003.
algo-1-s5jwv_1  | 2020-01-10 20:57:39,770 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycl

algo-1-s5jwv_1  | 2020-01-10 20:57:42,559 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Backend worker process die.
algo-1-s5jwv_1  | 2020-01-10 20:57:42,560 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Traceback (most recent call last):
algo-1-s5jwv_1  | 2020-01-10 20:57:42,561 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 172, in <module>
algo-1-s5jwv_1  | 2020-01-10 20:57:42,562 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     worker.run_server()
algo-1-s5jwv_1  | 2020-01-10 20:57:42,562 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 150, in run_server
algo-1-s5jwv_1  | 2020-01-10 20:57:42,563 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     self.handle_connection(cl_socke

algo-1-s5jwv_1  | 2020-01-10 20:57:42,818 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Backend worker process die.
algo-1-s5jwv_1  | 2020-01-10 20:57:42,818 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Traceback (most recent call last):
algo-1-s5jwv_1  | 2020-01-10 20:57:42,818 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 172, in <module>
algo-1-s5jwv_1  | 2020-01-10 20:57:42,818 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     worker.run_server()
algo-1-s5jwv_1  | 2020-01-10 20:57:42,818 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 150, in run_server
algo-1-s5jwv_1  | 2020-01-10 20:57:42,818 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     self.handle_connection(cl_socke

algo-1-s5jwv_1  | 2020-01-10 20:57:47,456 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Listening on port: /home/model-server/tmp/.mms.sock.9003
algo-1-s5jwv_1  | 2020-01-10 20:57:47,458 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - [PID]387
algo-1-s5jwv_1  | 2020-01-10 20:57:47,458 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - MXNet worker started.
algo-1-s5jwv_1  | 2020-01-10 20:57:47,458 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Python runtime: 3.6.8
algo-1-s5jwv_1  | 2020-01-10 20:57:47,458 [INFO ] W-9003-model com.amazonaws.ml.mms.wlm.WorkerThread - Connecting to: /home/model-server/tmp/.mms.sock.9003
algo-1-s5jwv_1  | 2020-01-10 20:57:47,459 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Connection accepted: /home/model-server/tmp/.mms.sock.9003.
algo-1-s5jwv_1  | 2020-01-10 20:57:47,682 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycl

algo-1-s5jwv_1  | 2020-01-10 20:57:50,485 [INFO ] W-9006-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Backend worker process die.
algo-1-s5jwv_1  | 2020-01-10 20:57:50,485 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Backend worker process die.
algo-1-s5jwv_1  | 2020-01-10 20:57:50,486 [INFO ] W-9006-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Traceback (most recent call last):
algo-1-s5jwv_1  | 2020-01-10 20:57:50,486 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Traceback (most recent call last):
algo-1-s5jwv_1  | 2020-01-10 20:57:50,486 [INFO ] W-9006-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 172, in <module>
algo-1-s5jwv_1  | 2020-01-10 20:57:50,486 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 172, in <module

algo-1-s5jwv_1  | 2020-01-10 20:57:50,737 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Backend worker process die.
algo-1-s5jwv_1  | 2020-01-10 20:57:50,737 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Traceback (most recent call last):
algo-1-s5jwv_1  | 2020-01-10 20:57:50,737 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 172, in <module>
algo-1-s5jwv_1  | 2020-01-10 20:57:50,737 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     worker.run_server()
algo-1-s5jwv_1  | 2020-01-10 20:57:50,737 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 150, in run_server
algo-1-s5jwv_1  | 2020-01-10 20:57:50,737 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     self.handle_connection(cl_socke

algo-1-s5jwv_1  | 2020-01-10 20:57:58,820 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Listening on port: /home/model-server/tmp/.mms.sock.9004
algo-1-s5jwv_1  | 2020-01-10 20:57:58,820 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - [PID]465
algo-1-s5jwv_1  | 2020-01-10 20:57:58,821 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - MXNet worker started.
algo-1-s5jwv_1  | 2020-01-10 20:57:58,821 [INFO ] W-9004-model com.amazonaws.ml.mms.wlm.WorkerThread - Connecting to: /home/model-server/tmp/.mms.sock.9004
algo-1-s5jwv_1  | 2020-01-10 20:57:58,821 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Python runtime: 3.6.8
algo-1-s5jwv_1  | 2020-01-10 20:57:58,822 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Connection accepted: /home/model-server/tmp/.mms.sock.9004.
algo-1-s5jwv_1  | 2020-01-10 20:57:58,931 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycl

algo-1-s5jwv_1  | 2020-01-10 20:58:01,570 [INFO ] W-9005-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Backend worker process die.
algo-1-s5jwv_1  | 2020-01-10 20:58:01,570 [INFO ] W-9005-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Traceback (most recent call last):
algo-1-s5jwv_1  | 2020-01-10 20:58:01,570 [INFO ] W-9005-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 172, in <module>
algo-1-s5jwv_1  | 2020-01-10 20:58:01,570 [INFO ] W-9005-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     worker.run_server()
algo-1-s5jwv_1  | 2020-01-10 20:58:01,570 [INFO ] W-9005-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 150, in run_server
algo-1-s5jwv_1  | 2020-01-10 20:58:01,570 [INFO ] W-9005-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     self.handle_connection(cl_socke

algo-1-s5jwv_1  | 2020-01-10 20:58:14,082 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Listening on port: /home/model-server/tmp/.mms.sock.9003
algo-1-s5jwv_1  | 2020-01-10 20:58:14,082 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - [PID]497
algo-1-s5jwv_1  | 2020-01-10 20:58:14,082 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - MXNet worker started.
algo-1-s5jwv_1  | 2020-01-10 20:58:14,082 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Python runtime: 3.6.8
algo-1-s5jwv_1  | 2020-01-10 20:58:14,082 [INFO ] W-9003-model com.amazonaws.ml.mms.wlm.WorkerThread - Connecting to: /home/model-server/tmp/.mms.sock.9003
algo-1-s5jwv_1  | 2020-01-10 20:58:14,083 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Connection accepted: /home/model-server/tmp/.mms.sock.9003.
algo-1-s5jwv_1  | 2020-01-10 20:58:14,502 [INFO ] W-9006-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycl

algo-1-s5jwv_1  | 2020-01-10 20:58:17,249 [INFO ] W-9000-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Backend worker process die.
algo-1-s5jwv_1  | 2020-01-10 20:58:17,250 [INFO ] W-9000-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Traceback (most recent call last):
algo-1-s5jwv_1  | 2020-01-10 20:58:17,250 [INFO ] W-9000-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 172, in <module>
algo-1-s5jwv_1  | 2020-01-10 20:58:17,250 [INFO ] epollEventLoopGroup-4-14 com.amazonaws.ml.mms.wlm.WorkerThread - 9000 Worker disconnected. WORKER_STARTED
algo-1-s5jwv_1  | 2020-01-10 20:58:17,250 [INFO ] W-9000-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     worker.run_server()
algo-1-s5jwv_1  | 2020-01-10 20:58:17,252 [INFO ] W-9000-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 150, in run

algo-1-s5jwv_1  | 2020-01-10 20:58:17,514 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Backend worker process die.
algo-1-s5jwv_1  | 2020-01-10 20:58:17,514 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Traceback (most recent call last):
algo-1-s5jwv_1  | 2020-01-10 20:58:17,514 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 172, in <module>
algo-1-s5jwv_1  | 2020-01-10 20:58:17,514 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     worker.run_server()
algo-1-s5jwv_1  | 2020-01-10 20:58:17,514 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 150, in run_server
algo-1-s5jwv_1  | 2020-01-10 20:58:17,514 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     self.handle_connection(cl_socke

algo-1-s5jwv_1  | 2020-01-10 20:58:21,038 [ERROR] Thread-2 com.amazonaws.ml.mms.metrics.MetricCollector - --- Logging error ---
algo-1-s5jwv_1  | Traceback (most recent call last):
algo-1-s5jwv_1  |   File "/usr/local/lib/python3.6/site-packages/psutil/_common.py", line 342, in wrapper
algo-1-s5jwv_1  |     ret = self._cache[fun]
algo-1-s5jwv_1  | AttributeError: _cache
algo-1-s5jwv_1  | 
algo-1-s5jwv_1  | During handling of the above exception, another exception occurred:
algo-1-s5jwv_1  | 
algo-1-s5jwv_1  | Traceback (most recent call last):
algo-1-s5jwv_1  |   File "/usr/local/lib/python3.6/site-packages/psutil/_pslinux.py", line 1513, in wrapper
algo-1-s5jwv_1  |     return fun(self, *args, **kwargs)
algo-1-s5jwv_1  |   File "/usr/local/lib/python3.6/site-packages/psutil/_common.py", line 345, in wrapper
algo-1-s5jwv_1  |     return fun(self)
algo-1-s5jwv_1  |   File "/usr/local/lib/python3.6/site-packages/psutil/_pslinux.py", line 1559, in _parse_stat_file
algo-1-s5jwv_1  |     wi

algo-1-s5jwv_1  | 2020-01-10 20:58:37,820 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Listening on port: /home/model-server/tmp/.mms.sock.9003
algo-1-s5jwv_1  | 2020-01-10 20:58:37,821 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - [PID]554
algo-1-s5jwv_1  | 2020-01-10 20:58:37,821 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - MXNet worker started.
algo-1-s5jwv_1  | 2020-01-10 20:58:37,821 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Python runtime: 3.6.8
algo-1-s5jwv_1  | 2020-01-10 20:58:37,821 [INFO ] W-9003-model com.amazonaws.ml.mms.wlm.WorkerThread - Connecting to: /home/model-server/tmp/.mms.sock.9003
algo-1-s5jwv_1  | 2020-01-10 20:58:37,822 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Connection accepted: /home/model-server/tmp/.mms.sock.9003.
algo-1-s5jwv_1  | 2020-01-10 20:58:38,391 [INFO ] W-9006-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycl

algo-1-s5jwv_1  | 2020-01-10 20:58:40,947 [INFO ] W-9001-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Backend worker process die.
algo-1-s5jwv_1  | 2020-01-10 20:58:40,948 [INFO ] W-9001-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Traceback (most recent call last):
algo-1-s5jwv_1  | 2020-01-10 20:58:40,949 [INFO ] W-9001-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 172, in <module>
algo-1-s5jwv_1  | 2020-01-10 20:58:40,950 [INFO ] W-9001-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     worker.run_server()
algo-1-s5jwv_1  | 2020-01-10 20:58:40,950 [INFO ] W-9001-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 150, in run_server
algo-1-s5jwv_1  | 2020-01-10 20:58:40,951 [INFO ] W-9001-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     self.handle_connection(cl_socke

algo-1-s5jwv_1  | 2020-01-10 20:58:41,203 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Backend worker process die.
algo-1-s5jwv_1  | 2020-01-10 20:58:41,204 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Traceback (most recent call last):
algo-1-s5jwv_1  | 2020-01-10 20:58:41,204 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 172, in <module>
algo-1-s5jwv_1  | 2020-01-10 20:58:41,204 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     worker.run_server()
algo-1-s5jwv_1  | 2020-01-10 20:58:41,204 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 150, in run_server
algo-1-s5jwv_1  | 2020-01-10 20:58:41,204 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     self.handle_connection(cl_socke

algo-1-s5jwv_1  | 2020-01-10 20:58:41,415 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Backend worker process die.
algo-1-s5jwv_1  | 2020-01-10 20:58:41,415 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Traceback (most recent call last):
algo-1-s5jwv_1  | 2020-01-10 20:58:41,415 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 172, in <module>
algo-1-s5jwv_1  | 2020-01-10 20:58:41,415 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     worker.run_server()
algo-1-s5jwv_1  | 2020-01-10 20:58:41,415 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 150, in run_server
algo-1-s5jwv_1  | 2020-01-10 20:58:41,416 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     self.handle_connection(cl_socke

algo-1-s5jwv_1  | 2020-01-10 20:59:15,578 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Listening on port: /home/model-server/tmp/.mms.sock.9007
algo-1-s5jwv_1  | 2020-01-10 20:59:15,579 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - [PID]640
algo-1-s5jwv_1  | 2020-01-10 20:59:15,580 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - MXNet worker started.
algo-1-s5jwv_1  | 2020-01-10 20:59:15,581 [INFO ] W-9007-model com.amazonaws.ml.mms.wlm.WorkerThread - Connecting to: /home/model-server/tmp/.mms.sock.9007
algo-1-s5jwv_1  | 2020-01-10 20:59:15,583 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Python runtime: 3.6.8
algo-1-s5jwv_1  | 2020-01-10 20:59:15,584 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Connection accepted: /home/model-server/tmp/.mms.sock.9007.
algo-1-s5jwv_1  | 2020-01-10 20:59:15,604 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycl

algo-1-s5jwv_1  | 2020-01-10 20:59:17,934 [INFO ] W-9006-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Backend worker process die.
algo-1-s5jwv_1  | 2020-01-10 20:59:17,934 [INFO ] W-9006-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Traceback (most recent call last):
algo-1-s5jwv_1  | 2020-01-10 20:59:17,935 [INFO ] W-9006-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 172, in <module>
algo-1-s5jwv_1  | 2020-01-10 20:59:17,935 [INFO ] W-9006-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     worker.run_server()
algo-1-s5jwv_1  | 2020-01-10 20:59:17,935 [INFO ] W-9006-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 150, in run_server
algo-1-s5jwv_1  | 2020-01-10 20:59:17,936 [INFO ] W-9006-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     self.handle_connection(cl_socke

algo-1-s5jwv_1  | 2020-01-10 20:59:18,220 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Backend worker process die.
algo-1-s5jwv_1  | 2020-01-10 20:59:18,220 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Traceback (most recent call last):
algo-1-s5jwv_1  | 2020-01-10 20:59:18,220 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 172, in <module>
algo-1-s5jwv_1  | 2020-01-10 20:59:18,220 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     worker.run_server()
algo-1-s5jwv_1  | 2020-01-10 20:59:18,220 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 150, in run_server
algo-1-s5jwv_1  | 2020-01-10 20:59:18,221 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     self.handle_connection(cl_socke

algo-1-s5jwv_1  | 2020-01-10 21:00:11,921 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Listening on port: /home/model-server/tmp/.mms.sock.9003
algo-1-s5jwv_1  | 2020-01-10 21:00:11,921 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - [PID]665
algo-1-s5jwv_1  | 2020-01-10 21:00:11,921 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - MXNet worker started.
algo-1-s5jwv_1  | 2020-01-10 21:00:11,921 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Python runtime: 3.6.8
algo-1-s5jwv_1  | 2020-01-10 21:00:11,921 [INFO ] W-9003-model com.amazonaws.ml.mms.wlm.WorkerThread - Connecting to: /home/model-server/tmp/.mms.sock.9003
algo-1-s5jwv_1  | 2020-01-10 21:00:11,922 [INFO ] W-9003-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Connection accepted: /home/model-server/tmp/.mms.sock.9003.
algo-1-s5jwv_1  | 2020-01-10 21:00:12,796 [INFO ] W-9005-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycl

algo-1-s5jwv_1  | 2020-01-10 21:00:15,293 [INFO ] W-9000-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Backend worker process die.
algo-1-s5jwv_1  | 2020-01-10 21:00:15,294 [INFO ] W-9000-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Traceback (most recent call last):
algo-1-s5jwv_1  | 2020-01-10 21:00:15,294 [INFO ] W-9000-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 172, in <module>
algo-1-s5jwv_1  | 2020-01-10 21:00:15,296 [INFO ] W-9000-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     worker.run_server()
algo-1-s5jwv_1  | 2020-01-10 21:00:15,296 [INFO ] W-9000-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 150, in run_server
algo-1-s5jwv_1  | 2020-01-10 21:00:15,297 [INFO ] W-9000-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     self.handle_connection(cl_socke

algo-1-s5jwv_1  | 2020-01-10 21:00:15,705 [INFO ] W-9001-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Backend worker process die.
algo-1-s5jwv_1  | 2020-01-10 21:00:15,705 [INFO ] W-9001-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Traceback (most recent call last):
algo-1-s5jwv_1  | 2020-01-10 21:00:15,705 [INFO ] W-9001-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 172, in <module>
algo-1-s5jwv_1  | 2020-01-10 21:00:15,705 [INFO ] W-9001-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     worker.run_server()
algo-1-s5jwv_1  | 2020-01-10 21:00:15,705 [INFO ] W-9001-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 150, in run_server
algo-1-s5jwv_1  | 2020-01-10 21:00:15,705 [INFO ] W-9001-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     self.handle_connection(cl_socke

algo-1-s5jwv_1  | 2020-01-10 21:00:15,969 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Backend worker process die.
algo-1-s5jwv_1  | 2020-01-10 21:00:15,970 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Traceback (most recent call last):
algo-1-s5jwv_1  | 2020-01-10 21:00:15,970 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 172, in <module>
algo-1-s5jwv_1  | 2020-01-10 21:00:15,970 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     worker.run_server()
algo-1-s5jwv_1  | 2020-01-10 21:00:15,970 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 150, in run_server
algo-1-s5jwv_1  | 2020-01-10 21:00:15,970 [INFO ] W-9004-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     self.handle_connection(cl_socke

algo-1-s5jwv_1  | 2020-01-10 21:01:45,031 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Listening on port: /home/model-server/tmp/.mms.sock.9007
algo-1-s5jwv_1  | 2020-01-10 21:01:45,032 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - [PID]756
algo-1-s5jwv_1  | 2020-01-10 21:01:45,033 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - MXNet worker started.
algo-1-s5jwv_1  | 2020-01-10 21:01:45,033 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Python runtime: 3.6.8
algo-1-s5jwv_1  | 2020-01-10 21:01:45,033 [INFO ] W-9007-model com.amazonaws.ml.mms.wlm.WorkerThread - Connecting to: /home/model-server/tmp/.mms.sock.9007
algo-1-s5jwv_1  | 2020-01-10 21:01:45,040 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Connection accepted: /home/model-server/tmp/.mms.sock.9007.
algo-1-s5jwv_1  | 2020-01-10 21:01:45,059 [INFO ] W-9002-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycl

algo-1-s5jwv_1  | 2020-01-10 21:01:47,419 [INFO ] W-9001-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Backend worker process die.
algo-1-s5jwv_1  | 2020-01-10 21:01:47,419 [INFO ] W-9001-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Traceback (most recent call last):
algo-1-s5jwv_1  | 2020-01-10 21:01:47,419 [INFO ] W-9001-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 172, in <module>
algo-1-s5jwv_1  | 2020-01-10 21:01:47,419 [INFO ] W-9001-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     worker.run_server()
algo-1-s5jwv_1  | 2020-01-10 21:01:47,419 [INFO ] W-9001-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 150, in run_server
algo-1-s5jwv_1  | 2020-01-10 21:01:47,419 [INFO ] W-9001-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     self.handle_connection(cl_socke

algo-1-s5jwv_1  | 2020-01-10 21:01:47,671 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Backend worker process die.
algo-1-s5jwv_1  | 2020-01-10 21:01:47,671 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle - Traceback (most recent call last):
algo-1-s5jwv_1  | 2020-01-10 21:01:47,671 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 172, in <module>
algo-1-s5jwv_1  | 2020-01-10 21:01:47,671 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     worker.run_server()
algo-1-s5jwv_1  | 2020-01-10 21:01:47,671 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -   File "/usr/local/lib/python3.6/site-packages/mms/model_service_worker.py", line 150, in run_server
algo-1-s5jwv_1  | 2020-01-10 21:01:47,671 [INFO ] W-9007-model-stdout com.amazonaws.ml.mms.wlm.WorkerLifeCycle -     self.handle_connection(cl_socke

### Clean Up

Remove the endpoint after we are done. 

In [ ]:
predictor.delete_endpoint()

## Side notes (Do NOT Run!)

In [ ]:
# !wget -qO- http://169.254.169.254/latest/meta-data/placement/availability-zone
# !aws ecr get-login --no-include-email --region us-east-1 --registry-ids 763104351884
# !docker pull 763104351884.dkr.ecr.us-east-1.amazonaws.com/mxnet-inference:1.6.0-gpu-py3


        (2, 
        ['[CLS]', 'where', 'did', 'super', 'bowl', '50', 'take', 'place', '?', 
        '[SEP]', 'super', 'bowl', '50', 'was', 'an', 'american', 'football', 
        'game', 'to', 'determine', 'the', 'champion', 'of', 'the', 'national', 
        'football', 'league', '(', 'nfl', ')', 'for', 'the', '2015', 'season', '.', 
        'the', 'american', 'football', 'conference', '(', 'afc', ')', 'champion', 
        'denver', 'broncos', 'defeated', 'the', 'national', 'football', 'conference', 
        '(', 'nfc', ')', 'champion', 'carolina', 'panthers', '24', '–', '10', 'to', 
        'earn', 'their', 'third', 'super', 'bowl', 'title', '.', 'the', 'game', 'was', 
        'played', 'on', 'february', '7', ',', '2016', ',', 'at', 'levi', "'", 's', 'stadium', 
        'in', 'the', 'san', 'francisco', 'bay', 'area', 'at', 'santa', 'clara', ',', 
        'california', '.', 'as', 'this', 'was', 'the', '50th', 'super', 'bowl', ',', 
        'the', 'league', 'emphasized', 'the', '"', 'golden', 'anniversary', '"', 'with', 
        'various', 'gold', '-', 'themed', 'initiatives', ',', 'as', 'well', 'as', 
        'temporarily', 'suspend', '##ing', 'the', 'tradition', 'of', 'naming', 
        'each', 'super', 'bowl', 'game', 'with', 'roman', 'nu', '##meral', '##s', 
        '(', 'under', 'which', 'the', 'game', 'would', 'have', 'been', 'known', 
        'as', '"', 'super', 'bowl', 'l', '"', ')', ',', 'so', 'that', 'the', 'logo', 
        'could', 'prominently', 'feature', 'the', 'arabic', 'nu', '##meral', '##s', 
        '50', '.', '[SEP]'], 
        
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], 
        
        168, 0, 0)